core/chains/evm/logpoller/log_poller.go

```diff
@@ -116,14 +117,104 @@ func NewLogPoller(orm *ORM, ec Client, lggr logger.Logger, pollPeriod time.Durat
 		backfillBatchSize: backfillBatchSize,
 		rpcBatchSize:      rpcBatchSize,
 		keepBlocksDepth:   keepBlocksDepth,
-		filters:           make(map[int]Filter),
+		filters:           make(map[string]Filter),
 		filterDirty:       true, // Always build filter on first call to cache an empty filter if nothing registered yet.
 	}
+
+	return lp
 }
 
+type AddressArray []common.Address
+type HashArray []common.Hash
+
 type Filter struct {
-	EventSigs []common.Hash
-	Addresses []common.Address
+	FilterName string
+	EventSigs  HashArray
+	Addresses  AddressArray
+}
+
+func (a *AddressArray) Scan(src interface{}) error {
+	baArray := pgtype.ByteaArray{}
+	err := baArray.Scan(src)
+	if err != nil {
+		return errors.New("Expected BYTEA[] column for AddressArray")
+	}
+	if baArray.Status != pgtype.Present || len(baArray.Dimensions) != 1 {
+		return errors.Errorf("Expected AddressArray to be 1-dimensional. Dimensions = %v", baArray.Dimensions)
+	}
+	addresses := []common.Address(*a)
+	lastAddress := common.Address{}
+	for i, ba := range baArray.Elements {
+		addr := common.Address{}
+		if ba.Status != pgtype.Present {
+			return errors.Errorf("Expected all addresses in AddressArray to be non-NULL.  Got AddressArray[%d] = NULL", i)
+		}
+		err = addr.Scan(ba.Bytes)
+		if err != nil {
+			return err
+		}
+		if addr != lastAddress { // dedup addresses
```

Comment:
I intended to guarantee they would be ordered, but look like there's a bug in my sql query... they have the wrong ORDER BY clause.  Thanks!

Reply:
Actually, I think this would all be a lot cleaner if I changed the db schema to have an array of events & topics on each row, rather than flattening them out to have one address,topic pair on each row.  Then there would be no need for deduping.

The only reason I avoided that, is in case in the future we might want to search by contract address or topic.  But maybe that's unlikely?

Reply:
Oh!  Found a better solution:  apparently you can use DISTINCT inside ARRAY_AGG; that removes the need for deduping in Scan()